### PR TITLE
Python: FunctionResultContent hash fix to handle lists/sets

### LINKED ISSUE #9977: Python: unhashable type 'list' exception in function_result_content __hash__ function
**Describe the bug**
A TypeError: unhashable type: 'list' exception is raised when invoking the hash function of `function_result_content`. 

**To Reproduce**
1. Create an agent group chat 
2. Add an agent to the group chat with a kernel function that returns a list
3. Invoke the agent through the group chat and let the agent invoke the kernel function that returns a list

**Expected behavior**
no exception if result of function_result_content is of an unhashable type

**Platform**
 - OS: [Windows]
 - IDE: [VS Code]
 - Language: [Python]
 - Source: [Python SDK 1.17.0]

**Additional context**
```
TypeError: unhashable type: 'list'
Traceback (most recent call last):
    async for content in group_chat.invoke():
  File "C:\Source\...\evaluation\.venv\Lib\site-packages\semantic_kernel\agents\group_chat\agent_group_chat.py", line 144, in invoke
    async for message in super().invoke_agent(selected_agent):
  File "C:\Source\....\evaluation\.venv\Lib\site-packages\semantic_kernel\agents\group_chat\agent_chat.py", line 144, in invoke_agent
    async for is_visible, message in channel.invoke(agent):
  File "C:\Source\....\evaluation\.venv\Lib\site-packages\semantic_kernel\agents\channels\chat_history_channel.py", line 75, in invoke
    mutated_history.add(mutated_message)
  File "C:\Source\....\evaluation\.venv\Lib\site-packages\semantic_kernel\contents\chat_message_content.py", line 318, in __hash__
    return hash((self.tag, self.role, self.content, self.encoding, self.finish_reason, *self.items))
  File "C:\Source\....\evaluation\.venv\Lib\site-packages\semantic_kernel\contents\function_result_content.py", line 190, in __hash__
    return hash((self.tag, self.id, self.result, self.name, self.function_name, self.plugin_name, self.encoding))
TypeError: unhashable type: 'list'
```

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

Currently, the hash function in `FunctionResultContent` does not properly handle types of list or set. Adding the proper handling to turn these into a tuple, and then the hash function works properly. 

### Description

This PR:
- Makes sure the FunctionResultContent `__hash__` method can properly handle mutable types like lists or sets
- Adds tests to ensure the new behavior works as expected
- Ensures the create message content method can handle lists or sets when generating the content and the model can only accept a string.
- Fixes #9977

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/agents/open_ai/assistant_content_generation.py [hunk 1]
```diff
@@ -87,17 +87,34 @@ def get_message_contents(message: "ChatMessageContent") -> list[dict[str, Any]]:
...
             case FileReferenceContent():
                 contents.append({
                     "type": "image_file",
                     "image_file": {"file_id": content.file_id},
                 })
+
             case FunctionResultContent():
-                contents.append({"type": "text", "text": content.result})
+                final_result = content.result
+                if isinstance(final_result, (list, tuple)):
```
Comment (eavanvalkenburg) [inline python/semantic_kernel/agents/open_ai/assistant_content_generation.py:112]: the `not` in the elif is throwing me off, maybe replace with a match case structure!
```suggestion
                match final_result:
                    case list() | tuple():
                        ...
                    case str():
                        ...
                     default:
                         final_result = str(final_result)
```

**Review by eavanvalkenburg — APPROVED** [review]
small note

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>semantic_kernel/agents/open_ai</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/agents/open_ai/assistant_content_generation.py">assistant_content_generation.py</a></td><td>153</td><td>12</td><td>92%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/agents/open_ai/assistant_content_generation.py#L93-L96">93&ndash;96</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/agents/open_ai/assistant_content_generation.py#L114-L115">114&ndash;115</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/agents/open_ai/assistant_content_generation.py#L348-L356">348&ndash;356</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/agents/open_ai/assistant_content_generation.py#L398">398</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/agents/open_ai/assistant_content_generation.py#L400">400</a></td></tr><tr><td colspan="5"><b>semantic_kernel/contents</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/contents/function_result_content.py">function_result_content.py</a></td><td>85</td><td>2</td><td>98%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/contents/function_result_content.py#L163-L165">163&ndash;165</a></td></tr><tr><td colspan="5"><b>semantic_kernel/functions</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/functions/kernel_function_from_prompt.py">kernel_function_from_prompt.py</a></td><td>154</td><td>7</td><td>95%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/functions/kernel_function_from_prompt.py#L167-L168">167&ndash;168</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/functions/kernel_function_from_prompt.py#L182">182</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/functions/kernel_function_from_prompt.py#L203">203</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/functions/kernel_function_from_prompt.py#L221">221</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/functions/kernel_function_from_prompt.py#L241">241</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/a2cddb4c8e0819ece53d39cbdd556782b5127ca4/semantic_kernel/functions/kernel_function_from_prompt.py#L324">324</a></td></tr><tr><td><b>TOTAL</b></td><td><b>16743</b></td><td><b>1860</b></td><td><b>89%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 2963 | 4 :zzz: | 0 :x: | 0 :fire: | 1m 41s :stopwatch: |

**Review by eavanvalkenburg — APPROVED** [review]

File: python/semantic_kernel/agents/open_ai/assistant_content_generation.py [hunk 2]
```diff
@@ -87,17 +87,38 @@ def get_message_contents(message: "ChatMessageContent") -> list[dict[str, Any]]:
...
+                final_result = content.result
+                match final_result:
+                    case str():
+                        contents.append({"type": "text", "text": final_result})
+                    case list() | tuple():
+                        contents.append({"type": "text", "text": " ".join(map(str, final_result))})
+                    case dict():
+                        contents.append({"type": "text", "text": str(final_result)})
+                    case _:
+                        contents.append({"type": "text", "text": str(final_result)})
```
Comment (TaoChenOSU) [inline python/semantic_kernel/agents/open_ai/assistant_content_generation.py:120]: nit: combine these two

**Review by TaoChenOSU — APPROVED** [review]

**Review by TaoChenOSU — APPROVED** [review]

**Pull request merged**